src/helpers.coffee

```diff
@@ -117,8 +117,8 @@ buildLocationData = (first, last) ->
 # Build a list of all comments attached to tokens.
 exports.extractAllCommentTokens = (tokens) ->
   allCommentsObj = {}
-  for token in tokens when token.comments
-    for comment in token.comments
+  for token in tokens when token.comments or token.prevToken?.comments
+    for comment in [(token.prevToken?.comments ? [])..., (token.comments ? [])...]
```

Comment:
Why is this change necessary? It feels like a change that may have gotten the particular test to pass, but in a larger codebase it might introduce bugs.

In particular, there’s logic in the comments output code that avoids outputting a comment if that comment has already been output. That logic is there because sometimes the same comment gets attached to multiple tokens, such as when tokens are duplicated in the rewriter. So what I’m concerned about is that adding a condition here for looking for comments in not just the current token but also in the previous one too makes _every_ case of comments attached to tokens this “duplicated comments” case, that I’m working around by checking that we don’t double-output comments; but that’s really suboptimal because ideally comments wouldn’t be duplicated in the first place. Like if you removed the “don’t double-output comments” code, I bet that every comment would be double-output now, not just the ones where we expect that problem because of rewriting.

In other words, perhaps something needs to be done in the rewriter to move/copy the comments to/from the semicolon token as appropriate, rather than this global change looking for comments beyond where we “should” be looking.

Reply:
Ok I updated to avoid the need for this change by instead using `moveComments()` to shift the comment (that was getting lost) from the semicolon token to the outdent token

Which then also gets that comment to output during compile-to-JS (which it wasn't previously, eg this comment gets lost on `master`:
```
if a
  ;
  ### comment ###
```

Reply:
Good! Thank you, that’s much better.

Doesn’t seem to fix #5241, but it’s on the right track.